server/src/main/java/org/apache/iotdb/db/query/reader/seriesRelated/AbstractDataReader.java

```diff
@@ -0,0 +1,515 @@
+    for (ChunkMetaData data : currentChunkMetaDataList) {
+      if (data.getChunkLoader() == null) {
+        data.setChunkLoader(
+            new ChunkLoaderImpl(new TsFileSequenceReader(resource.getFile().getAbsolutePath())));
+      }
+    }
+    if (!pathModifications.isEmpty()) {
+      QueryUtils.modifyChunkMetaData(currentChunkMetaDataList, pathModifications);
+    }
+    ReadOnlyMemChunk readOnlyMemChunk = resource.getReadOnlyMemChunk();
+    if (readOnlyMemChunk != null) {
+      currentChunkMetaDataList.add(readOnlyMemChunk.getChunkMetaData());
+    }
+    return currentChunkMetaDataList;
+  }
+
+  private TreeSet<TsFileResource> sortUnSeqFileResources(List<TsFileResource> tsFileResources) {
+    TreeSet<TsFileResource> unseqTsFilesSet = new TreeSet<>((o1, o2) -> {
+      Map<String, Long> startTimeMap = o1.getStartTimeMap();
+      Long minTimeOfO1 = startTimeMap.get(seriesPath.getDevice());
+      Map<String, Long> startTimeMap2 = o2.getStartTimeMap();
+      Long minTimeOfO2 = startTimeMap2.get(seriesPath.getDevice());
+
+      return Long.compare(minTimeOfO1, minTimeOfO2);
+    });
+    unseqTsFilesSet.addAll(tsFileResources);
+    return unseqTsFilesSet;
+  }
+
+  @Override
+  public boolean isManagedByQueryManager() {
+    return managedByQueryManager;
+  }
+
+  @Override
+  public void setManagedByQueryManager(boolean managedByQueryManager) {
+    this.managedByQueryManager = managedByQueryManager;
+  }
+
+  @Override
+  public boolean hasRemaining() {
+    return hasRemaining;
+  }
+
+  @Override
+  public void setHasRemaining(boolean hasRemaining) {
+    this.hasRemaining = hasRemaining;
+  }
+
+  /**
+   * Because you get a list of all the files, some files are not necessary when filters exist. This
+   * method filters out the available data files based on the filter
+   */
+  private void removeInvalidFiles() {
+    //filter seq files
+    while (filter != null && !seqFileResource.isEmpty()) {
+      if (!isValid(seqFileResource.get(0))) {
+        seqFileResource.remove(0);
+        continue;
+      }
+      break;
+    }
+    //filter unseq files
+    while (filter != null && !unseqFileResource.isEmpty()) {
+      if (!isValid(unseqFileResource.first())) {
+        unseqFileResource.pollFirst();
+        continue;
+      }
+      break;
+    }
+  }
+
+  private boolean isValid(TsFileResource tsFileResource) {
+    long startTime = tsFileResource.getStartTimeMap()
+        .getOrDefault(seriesPath.getDevice(), Long.MIN_VALUE);
+    long endTime = tsFileResource.getEndTimeMap()
+        .getOrDefault(seriesPath.getDevice(), Long.MAX_VALUE);
+    return filter.satisfyStartEndTime(startTime, endTime);
+  }
+
+  /**
+   * unseq file is a very special file that intersects not only with an ordered file, but also with
+   * another unseq file. So we need a way to find all the files that might be used to intersect the
+   * current measurement point.
```

Comment:
unseq files are very special files that intersect not only with sequence files, but also with
 other unseq files. So we need to find all tsfiles that overlapped with current chunk

Reply:
done